core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/simplify.kt

```diff
@@ -0,0 +1,84 @@
+package org.jetbrains.kotlinx.dataframe.api
+
+import org.jetbrains.kotlinx.dataframe.DataColumn
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.columns.ColumnSet
+import org.jetbrains.kotlinx.dataframe.columns.ColumnsResolver
+import org.jetbrains.kotlinx.dataframe.documentation.Indent
+import org.jetbrains.kotlinx.dataframe.documentation.LineBreak
+import org.jetbrains.kotlinx.dataframe.documentation.UsageTemplateColumnsSelectionDsl.UsageTemplate
+import org.jetbrains.kotlinx.dataframe.impl.columns.simplify
+import org.jetbrains.kotlinx.dataframe.impl.columns.transform
+import org.jetbrains.kotlinx.dataframe.util.TOP_MESSAGE
+
+// region ColumnsSelectionDsl
+
+/**
+ * See [Usage]
+ */
+public interface SimplifyColumnsSelectionDsl {
+
+    /**
+     * ## Simplify [ColumnSet] Usage
+     *
+     * @include [UsageTemplate]
+     * {@setArg [UsageTemplate.DefinitionsArg]
+     *  {@include [UsageTemplate.ColumnSetDef]}
+     * }
+     *
+     * {@setArg [UsageTemplate.ColumnSetFunctionsArg]
+     *  {@include [Indent]}{@include [ColumnSetName]}**`()`**
+     * }
+     *
+     * {@setArg [UsageTemplate.PlainDslPart]}
+     * {@setArg [UsageTemplate.ColumnGroupPart]}
+     */
+    public interface Usage {
+
+        /** .[**simplify**][ColumnsSelectionDsl.simplify] */
+        public interface ColumnSetName
+    }
+
+    /**
+     * ## Simplify [ColumnSet]
+     *
+     * Given a [ColumnSet], [simplify] simplifies the structure by removing columns that are already present in
+     * column groups in [this], returning only these groups plus columns not belonging in any of the groups.
+     *
+     * In other words, this means that if a column in [this] is inside another column group in [this],
+     * it will not be included in the result.
+     *
+     * See [Usage] for how to use [simplify].
+     *
+     * ## For example:
+     * 
+     * [cols][ColumnsSelectionDsl.cols]`(a, a.b, d.c).`[simplify][SimplifyColumnsSelectionDsl.simplify]`() == `[cols][ColumnsSelectionDsl.cols]`(a, d.c)`
+     * {@include [LineBreak]}
+     * `df.`[select][DataFrame.select]` { `[colsAtAnyDepth][ColumnsSelectionDsl.colsAtAnyDepth]` { "e" `[in][String.contains]` it.`[name][DataColumn.name]` }.`[simplify][ColumnSet.simplify]`() }`
+     *
+     * @return A [ColumnSet][ColumnSet]`<`[C][C]`>` containing only the columns that are not inside any column group in [this].
+     */
+    @Suppress("UNCHECKED_CAST")
+    public fun <C> ColumnSet<C>.simplify(): ColumnSet<C> = simplifyInternal() as ColumnSet<C>
+
+    // region deprecated
+
+    @Deprecated(TOP_MESSAGE, ReplaceWith("simplify()"), DeprecationLevel.WARNING)
```

Comment:
Need to say, that roots and top lead better to understanding what is doing this function? And it does not simplify my life, because semantic closer to (cut hierarchy or leave only high-level groups and leaves).

Also, for me, it's unclear, how it should work with 3 and more levels of hierarchy? Or could I apply on different level of our ColumnTree?

Reply:
Probably you could cover it with more interesting tests with 3 and more levels of hierarchy or change the functionality (like you add ignoreCase flag earlier, but with simplify(level=2))

Reply:
You cannot call it at a lower level of the column tree. It can only be called on a `ColumnSet` (so a result of `colsOf {}`, `colsAtAnyDepth {}`, `(a and b)`, etc.).

I didn't like the name `top` because that was totally non-descriptive. `roots` was better, but with that name one might expect `(a.b.c and a.b).roots()` to return the actual "root", which would be `a`, but it actually returns `a.b`...
I'm definitely open for other suggestions tho! Simplify works, but I also think it's too broad of a name for what it does.

Reply:
Adding a level does not really make sense to me in the context of this function. All this function does is remove columns that already appear in groups (in groups in groups...) in the column set. It's already a niche function that's used in the internals of the library a lot and just happens to be in the public api.
I cannot think of a use-case that adding a level would solve.